rest_framework_json_api/serializers.py

```diff
@@ -163,3 +167,140 @@ def get_field_names(self, declared_fields, info):
+    def get_polymorphic_model_for_serializer(cls, serializer):
+        """
+        Return the polymorphic model associated with the given serializer.
+        Raise `NotImplementedError` if no model is found for the given serializer. This usually
+        means that a serializer is missing in the class's `polymorphic_serializers` attribute.
+        """
+        try:
+            return cls._poly_serializer_model_map[serializer]
+        except KeyError:
+            raise NotImplementedError(
+                "No polymorphic model has been found for serializer {}".format(serializer.__name__))
+
+    @classmethod
+    def get_polymorphic_serializer_for_type(cls, obj_type):
+        """
+        Return the polymorphic serializer associated with the given type.
+        Raise `NotImplementedError` if no serializer is found for the given type. This usually
+        means that a serializer is missing in the class's `polymorphic_serializers` attribute.
+        """
+        try:
+            return cls._poly_type_serializer_map[obj_type]
+        except KeyError:
+            raise NotImplementedError(
+                "No polymorphic serializer has been found for type {}".format(obj_type))
+
+    @classmethod
+    def get_polymorphic_model_for_type(cls, obj_type):
+        """
+        Return the polymorphic model associated with the given type.
+        Raise `NotImplementedError` if no model is found for the given type. This usually
+        means that a serializer is missing in the class's `polymorphic_serializers` attribute.
+        """
+        return cls.get_polymorphic_model_for_serializer(
+            cls.get_polymorphic_serializer_for_type(obj_type))
+
+    @classmethod
+    def get_polymorphic_types(cls):
+        """
+        Return the list of accepted types.
+        """
+        return cls._poly_type_serializer_map.keys()
+
+    def to_representation(self, instance):
+        """
+        Retrieve the appropriate polymorphic serializer and use this to handle representation.
+        """
+        serializer_class = self.get_polymorphic_serializer_for_instance(instance)
+        return serializer_class(instance, context=self.context).to_representation(instance)
+
+    def to_internal_value(self, data):
+        """
+        Ensure that the given type is one of the expected polymorphic types, then retrieve the
+        appropriate polymorphic serializer and use this to handle internal value.
+        """
+        received_type = data.get('type')
+        expected_types = self.get_polymorphic_types()
+        if received_type not in expected_types:
+            raise Conflict(
+                'Incorrect relation type. Expected on of [{expected_types}], '
+                'received {received_type}.'.format(
+                    expected_types=', '.join(expected_types), received_type=received_type))
+        serializer_class = self.get_polymorphic_serializer_for_type(received_type)
+        self.__class__ = serializer_class
```

Comment:
That's a trick to replace the current instance class (the polymorphic "parent") by the correct subclass without altering the dynamic data that may have been associated with the instance. It is necessary to get the correct defaults (such as subclass Meta, custom validators) for the rest of the process.

I'm not very proud of that too, I'll search how this case may have been handled by other drf libraries and I'm open to any suggestions.